src/main/scala/com/mozilla/telemetry/stats/MannWhitneyU.scala

```diff
@@ -0,0 +1,120 @@
+/*
+ This Source Code Form is subject to the terms of the Mozilla Public
+ License, v. 2.0. If a copy of the MPL was not distributed with this
+ file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package com.mozilla.telemetry.stats
+
+
+// Constants
+
+object MannWhitneyU {
+
+  def erf(x: Double): Double =  {
+    // Constants.
+    val a1: Double = 0.254829592
+    val a2: Double = -0.284496736
+    val a3: Double = 1.421413741
+    val a4: Double = -1.453152027
+    val a5: Double = 1.061405429
+    val p: Double = 0.3275911
+
+    if (x == 0) {
+      return 0.0
+    }
+
+    val sign: Int = if (x < 0) -1 else 1
+    val absx: Double = math.abs(x)
+    val t: Double = 1.0 / (1.0 + p * absx);
+    val y: Double = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * math.exp(-x * x);
+    sign * y
+  }
+
+  def erfc(x: Double): Double = {
+    1.0 - erf(x)
+  }
+
+  def ndtr(a: Double): Double = {
+    /*
+    Returns the area under the Gaussian probability density function,
+    integrated from minus infinity to x.
+
+    See: https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.ndtr.html#scipy.special.ndtr
+
+    */
+    val sqrth: Double = math.sqrt(2) / 2
+    val x: Double = a * sqrth
+    val z: Double = math.abs(x)
+    var y: Double = 0.0
+
+    if (z < sqrth) {
+        y = 0.5 + 0.5 * erf(x)
+    } else {
+        y = 0.5 * erfc(z)
+        if (x > 0) {
+            y = 1.0 - y
+        }
+    }
+    return y
+  }
+
+  def tie_correct(sample: Map[Int, Int]): Double = {
+    var tc: Double = 0.0
+    val n = sample.foldLeft[Int](0)(_ + _._2)
+
+    if (n < 2) {
+      return 1.0
+    }
+
+    tc = sample.foldLeft[Double](tc)((sum, i) => sum + math.pow(i._2, 3) - i._2)
+    tc = 1 - tc / (math.pow(n, 3) - n)
+    tc
+  }
+
+  def rank(sample: Map[Int, Int]): Map[Int, Double] = {
+    var rank: Int = 1
+    val ranks = scala.collection.mutable.Map[Int, Double]()
+    val sorted = scala.collection.immutable.SortedMap[Int, Int]() ++ sample
+
+    sorted.foreach {
+      case (k, v) => {
+        ranks(k) = rank + (v.toDouble - 1) / 2
+        rank += v
+      }
+    }
+
+    ranks.toMap
+  }
+
+  def mwu(sample1: Map[Int, Int], sample2: Map[Int, Int],
+          use_continuity: Boolean = true) = {
+    // Merge maps, adding values if keys match.
+    val sample = sample1 ++ sample2.map {
+      case (k,v) => {
+        k -> (v + sample1.getOrElse(k, 0))
+      }
+    }
+
+    val ranks = rank(sample)
+    val sum_of_ranks = sample1.foldLeft[Double](0.0)(
+      (sum, i) => sum + (i._2 * ranks(i._1)))
+    val n1 = sample1.foldLeft[Int](0)(_ + _._2)
+    val n2 = sample2.foldLeft[Int](0)(_ + _._2)
+
+    // Calculate Mann-Whitney U for both samples.
+    val u1 = sum_of_ranks - (n1 * (n1 + 1)) / 2
+    val u2 = n1 * n2 - u1
+
+    val tc = tie_correct(sample)
+    // TODO: If zero raise an error
```

Comment:
I didn't know the proper way to handle this in Scala.

Reply:
Here's an example: https://github.com/mozilla/telemetry-streaming/blob/db4c97da1357000436318fe51e7b9cc509ba11f4/src/main/scala/com/mozilla/telemetry/timeseries/package.scala#L38